File: src/python/maps/map10.py
import ika
from soulreaver import SoulReaver
from thing import Thing
from rune import FireRune
import sound

def AutoExec(engineRef):
    if 'fireguard' not in engineRef.saveFlags:
        engineRef.mapThings.append(RuneListener(engineRef))

    if 'firerune' in engineRef.saveFlags:
        ika.Map.RemoveEntity(ika.Map.entities['demiyeti'])
    else:
        engineRef.mapThings.append(DeathListener(engineRef))

def to9(engineRef):
    yield from engineRef.mapSwitchTask('map09.ika-map', (engineRef.player.x, 1 * 16))

class DeathListener(Thing):
    'Waits until the yeti is dead, then drops the fire rune.'
    def __init__(self, engineRef, yeti=None):
        self.engineRef = engineRef
        self.yeti = yeti

    def update(self):
        if not self.yeti:
            # have to get the entity here, since it hasn't been created yet
            # in AutoExec. (if we had more time, I'd fix that problem instead of
            # doing this)
            sound.playMusic("music/Competative.xm")
            self.yeti = self.engineRef.entFromEnt[
                ika.Map.entities['demiyeti'].name
                ]
        elif self.yeti.stats.hp == 0:
            if 'nearend' not in self.engineRef.saveFlags:
                e = ika.Entity(71, 132, 2, 'firerune.ika-sprite')
                e.name = 'firerune'
                self.engineRef.addEntity(
                    FireRune(e)
                    )
            else:
                sound.playMusic("music/winter.ogg")
                self.engineRef.saveFlags['fireguard'] = 'True'

            return True

    def draw(self):
        pass

class RuneListener(object):
    def __init__(self, engineRef):
        self.engineRef = engineRef

    def update(self):
        if 'nearend' in self.engineRef.saveFlags:
            sound.playMusic('music/resurrection.it')
            y = SoulReaver(ika.Entity(21*16, 13*16, 2, 'soulreaver.ika-sprite'))
            self.engineRef.addEntity(y)
            self.engineRef.mapThings.append(DeathListener(self.engineRef, y))
            return True

    def draw(self):
        pass

File: src/winter/maps/map10.ts
import { MapScript } from "./mapscript.js"
import { FireRune } from "./../rune.js"
import { Thing } from "./../thing.js"
import { PyEngine } from "./../winter.js"
import { SoulReaver, Yeti } from "./../yeti.js"

export default new MapScript(autoexec, { to9 })

export function autoexec(engineRef: PyEngine) {
    const engine = engineRef.getEngine()
    if (!engineRef.hasSaveFlag('fireguard')) {
        engineRef.addMapThing(new RuneListener(engineRef))
    }

    if (engineRef.hasSaveFlag('firerune')) {
        engine.map.removeSprite(engine.map.sprites['demiyeti'])
    } else {
        engineRef.addMapThing(new DeathListener(engineRef))
    }
}

export function *to9(engineRef: PyEngine) {
    const p = engineRef.getPlayerEntity()
    yield* engineRef.mapSwitchTask('map09.ika-map', [p.sprite.x, 1 * 16])
}

class DeathListener extends Thing {
    // Waits until the yeti is dead, then drops the fire rune.
    constructor(private engineRef: PyEngine, private yeti: Yeti | null = null) { super() }

    update() {
        const engine = this.engineRef.getEngine()
        if (this.yeti === null) {
            // have to get the entity here, since it hasn't been created yet
            // in AutoExec. (if we had more time, I'd fix that problem instead of
            // doing this)
            this.engineRef.playMusic("music/Competative.xm")
            const y = this.engineRef.getEntityForSpriteName(engine.map.sprites['demiyeti'].name)
            if (!(y instanceof Yeti)) { throw new Error("Expected Yeti") }
            this.yeti = y
        } else if (this.yeti.stats.hp === 0) {
            if (!this.engineRef.hasSaveFlag('nearend')) {
                const e = engine.map.addSprite(71, 132, 2, 'firerune.ika-sprite')
                e.name = 'firerune'
                this.engineRef.addEntity(new FireRune(this.engineRef, e))
            } else {
                this.engineRef.playMusic("music/winter.ogg")
                this.engineRef.setSaveFlag('fireguard', 'True')
            }
            return true
        }
        return false
    }
}

class RuneListener extends Thing {
    constructor(private engineRef: PyEngine) { super() }

    update() {
        const engine = this.engineRef.getEngine()
        if (this.engineRef.hasSaveFlag('nearend')) {
            this.engineRef.playMusic('music/resurrection.it')
            const y = new SoulReaver(this.engineRef, engine.map.addSprite(21*16, 13*16, 2, 'soulreaver.ika-sprite'))
            this.engineRef.addEntity(y)
            this.engineRef.addMapThing(new DeathListener(this.engineRef, y))
            return true
        }
        return false
    }
}
